File: GothicCalculatorApi/Models/LockModels.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace GothicCalculatorApi.Models;

public sealed class LockDefinitionRequest
{
    public string Name { get; set; } = string.Empty;

    [MinLength(2)]
    [MaxLength(8)]
    public List<string> Rules { get; set; } = [];

    [MinLength(2)]
    [MaxLength(8)]
    public List<int> Start { get; set; } = [];
}

public sealed class LockInstruction
{
    public int Plate { get; set; }
    public int Count { get; set; }
    public string Direction { get; set; } = string.Empty;
}

public sealed class LockSolveResponse
{
    public bool Ok { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Name { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Status { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Lines { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? Steps { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<LockInstruction>? Instructions { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Error { get; set; }
}

public sealed class HealthResponse
{
    public string Status { get; set; } = "ok";
    public bool BinaryFound { get; set; }
    public string BinaryPath { get; set; } = string.Empty;
}

File: gothic-calculator-frontend/src/app/models/lock.models.ts
export interface LockDefinition {
  name: string;
  rules: string[];
  start: number[];
}

export interface LockInstruction {
  plate: number;
  count: number;
  direction: 'left' | 'right';
}

export interface LockSolveResponse {
  ok: boolean;
  name?: string;
  status?: 'solved' | 'already_open';
  lines?: number;
  steps?: number;
  instructions?: LockInstruction[];
  error?: string;
}

export const DEFAULT_LOCK_NAME = 'my lock';

export const TOWER_CHEST_TEMPLATE: LockDefinition = {
  name: DEFAULT_LOCK_NAME,
  rules: ['3r, 6l', '-', '1r, 4l, 6r', '2r, 5r, 6l', '-', '3l'],
  start: [5, 3, 6, 7, 2, 7],
};

export const MIN_PLATES = 2;
export const MAX_PLATES = 8;
export const MIN_POSITION = 1;
export const MAX_POSITION = 7;
export const GOAL_POSITION = 4;
